### PR TITLE
Add HuggingFacePipeline LLM

### PR DESCRIPTION
**Pull request by sam-h-bean**
https://github.com/hwchase17/langchain/issues/354

Add support for running your own HF pipeline locally. This would allow you to get a lot more dynamic with what HF features and models you support since you wouldn't be beholden to what is hosted in HF hub. You could also do stuff with HF Optimum to quantize your models and stuff to get pretty fast inference even running on a laptop.

### CODE REVIEW

**Comment by hwchase17** [comment]
> #354
> 
> Add support for running your own HF pipeline locally. This would allow you to get a lot more dynamic with what HF features and models you support since you wouldn't be beholden to what is hosted in HF hub. You could also do stuff with HF Optimum to quantize your models and stuff to get pretty fast inference even running on a laptop.
> 
> Let me know if you want this code and I will clean it up.

want it? i need it!

but in all seriousness, i love it, and i know others will as well (has been a common ask) - thank you!!!

**Comment by sam-h-bean** [comment]
@hwchase17 Last issue I am having is that the integration tests now require torch to be installed for HF pipelines to run. I am trying to add it to poetry but I might do it incorrectly. Any examples of that I can look at?

**Comment by sam-h-bean** [comment]
So I added torch to the pyproject.yaml and am still seeing errors for torch missing in the poetry environment. Not really sure if I'm missing something or not.
```shell
test_huggingface_pipeline.py::test_huggingface_pipeline_text_generation FAILED [100%]None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.


```

File: langchain/llms/huggingface_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            **{"model_kwargs": _model_kwargs},
+        }
+
+    @property
+    def _llm_type(self) -> str:
+        return "huggingface_local"
+
+    def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
+        response = self.pipeline(text_inputs=prompt)
+        if "error" in response:
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:87]: can this still happen with the local pipeline? probably not right? not sure we need this check

Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:87]: done

File: pyproject.toml [hunk 2]
```diff
@@ -22,6 +22,7 @@ spacy = {version = "^3", optional = true}
 nltk = {version = "^3", optional = true}
 transformers = {version = "^4", optional = true}
 beautifulsoup4 = {version = "^4", optional = true}
+torch = {version = "^1.13.1", optional = true}
```
[the remark this answers is not in the document]
Comment (sam-h-bean) [reply pyproject.toml:25]: done

File: langchain/llms/huggingface_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """Wrapper around HuggingFace Pipeline API.
+
+    To use, you should have the ``transformers`` python package installed.
+
+    Only supports `text-generation` and `text2text-generation` for now.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import HuggingFacePipeline
```
[the remark this answers is not in the document]
Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:23]: done

File: langchain/llms/huggingface_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+"""Wrapper around HuggingFace APIs."""
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:1]: ubernit: this looks copied/same as huggingface hub, can we maybe say "huggingface pipelines" or something?

File: langchain/llms/huggingface_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        """Validate python package exists in environment."""
+        try:
+            from transformers import AutoModelForCausalLM, AutoTokenizer
+            from transformers import pipeline as hf_pipeline
+
+            model_id = values["model_id"]
+            tokenizer = AutoTokenizer.from_pretrained(model_id)
+            model = AutoModelForCausalLM.from_pretrained(model_id)
+
+            pipeline = hf_pipeline(
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:52]: honestly asking: would it be nice to just let users pass in a pipeline directly?

if so, what we can do is make a classmethod `from_model_id` which does this logic, and then passes the pipeline into init

that way we can have (1) the flexibility of passing in an arbitrary pipeline, (2) an easy init with model_id (just in differnet places)

Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:52]: probably

Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:52]: One question @hwchase17 how do I pass into init with pydantic?

File: langchain/llms/huggingface_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate python package exists in environment."""
+        try:
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:44]: i would probably put the try-except around the imports in the from_model_id method

also, this isnt exactly what i meant by class method - ill comment below

File: langchain/llms/huggingface_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        except ImportError:
+            raise ValueError(
+                "Could not import transformers python package. "
+                "Please it install it with `pip install transformers`."
+            )
+
+        return values
+
+    @staticmethod
+    def from_model_id(values):
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:57]: i would do something like
```
@classmethod
def from_model_id(model_id: str, task: Optional[str] = None, model_kwargs: Optional[dict] = None, **kwargs):
    # construct pipeline here
    return cls(pipeline=pipeline, **kwargs)
```

this allows users to do either

`HuggingFacePipeline(pipeline=pipeline)` or `HuggingFacePipeline.from_model_id("gpt2")`

i think if we do this, we can also just delete the above validate_environment call

i think we also no longer need to save the model_id, model_kwargs, and task param on the class (which simplifies the interface imensely)

Comment (hwchase17) [reply langchain/llms/huggingface_pipeline.py:57]: _identifying_params probably becomes something just based on the pipeline - what would that look like? we use that for caching...

File: langchain/llms/huggingface_pipeline.py (hunk 6 above)
Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:44]: moved to inner method

File: langchain/llms/huggingface_pipeline.py (hunk 7 above)
Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py]: Moved to classmethod. Will think about changing identifying params now

Comment (hwchase17) [reply langchain/llms/huggingface_pipeline.py]: makes sense, i think thats reasonable

File: langchain/llms/huggingface_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            pipeline = hf_pipeline(
+                task=task, model=model, tokenizer=tokenizer, **model_kwargs
+            )
+            if pipeline.task not in VALID_TASKS:
+                raise ValueError(
+                    f"Got invalid task {pipeline.task}, "
+                    f"currently only {VALID_TASKS} are supported"
+                )
+
+            return cls(pipeline=pipeline, **kwargs)
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:60]: we should probably also pass in model_id and model_kwargs now that we're keeping them around

File: langchain/llms/huggingface_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        .. code-block:: python
+
+            from langchain.llms.huggingface_pipeline import HuggingFacePipeline
+            hf = HuggingFacePipeline(model_id="gpt2", task="text-generation")
+    """
+
+    pipeline: Any  #: :meta private:
+    model_id: str = DEFAULT_MODEL_ID
+    """Model name to use."""
+    task: str = DEFAULT_TASK
```
Comment (hwchase17) [inline langchain/llms/huggingface_pipeline.py:31]: i dont think this is ever used (since we access it from the pipeline)

**Review by hwchase17 — COMMENTED** [review]
some final nits but other than that lgtm!

File: langchain/llms/huggingface_pipeline.py (hunk 9 above)
Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py:31]: removed

File: langchain/llms/huggingface_pipeline.py (hunk 8 above)
Comment (sam-h-bean) [reply langchain/llms/huggingface_pipeline.py]: done!

**Comment by sam-h-bean** [comment]
@hwchase17 One last time!

File: langchain/llms/huggingface_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        extra = Extra.forbid
+
+    @classmethod
+    def from_model_id(
+        cls,
+        model_id: str,
+        task: str,
+        model_kwargs: Optional[dict] = None,
+        **kwargs: Any,
+    ) -> LLM:
```
Comment (sam-h-bean) [inline langchain/llms/huggingface_pipeline.py:47]: I could change this to `langchain.HuggingFacePipeline`, it wouldn't let me do just `HuggingFacePipeline`... Not sure why. Let me know if you want this to change

Comment (hwchase17) [reply langchain/llms/huggingface_pipeline.py:47]: no i think this is fine!

**Review by hwchase17 — APPROVED** [review]

**Comment by hwchase17** [comment]
ah very close! just some lint. lmk if you want me to tackle

hoping to get this in release tmrw. whats your twitter? want to give you a big shoutout

**Comment by sam-h-bean** [comment]
@hwchase17 got it. take a look.

I have no twitter but definitely shout out my company's socials https://twitter.com/YouSearchEngine

**Pull request merged**